subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java

```diff
@@ -32,82 +34,99 @@
+
+    public ZipHasher(ResourceHasher resourceHasher, ResourceFilter resourceFilter) {
+        this.resourceHasher = resourceHasher;
+        this.resourceFilter = resourceFilter;
     }
 
     @Nullable
     @Override
     public HashCode hash(RegularFileSnapshot fileSnapshot) {
-        return hashJarContents(fileSnapshot);
+        return hashZipContents(fileSnapshot);
     }
 
     @Override
     public void appendConfigurationToHasher(Hasher hasher) {
         hasher.putString(getClass().getName());
-        classpathResourceHasher.appendConfigurationToHasher(hasher);
-        classpathResourceFilter.appendConfigurationToHasher(hasher);
+        resourceHasher.appendConfigurationToHasher(hasher);
+        resourceFilter.appendConfigurationToHasher(hasher);
     }
 
     @Nullable
-    private HashCode hashJarContents(RegularFileSnapshot jarFileSnapshot) {
+    private HashCode hashZipContents(RegularFileSnapshot zipFileSnapshot) {
         try {
-            List<FileSystemLocationFingerprint> fingerprints = fingerprintZipEntries(jarFileSnapshot.getAbsolutePath());
+            List<FileSystemLocationFingerprint> fingerprints = fingerprintZipEntries(zipFileSnapshot.getAbsolutePath());
             if (fingerprints.isEmpty()) {
                 return null;
             }
             Hasher hasher = Hashing.newHasher();
             NormalizedPathFingerprintCompareStrategy.appendSortedToHasher(hasher, fingerprints);
             return hasher.hash();
         } catch (Exception e) {
-            return hashMalformedZip(jarFileSnapshot, e);
+            return hashMalformedZip(zipFileSnapshot, e);
         }
     }
 
-    private List<FileSystemLocationFingerprint> fingerprintZipEntries(String jarFile) throws IOException {
-        List<FileSystemLocationFingerprint> fingerprints = Lists.newArrayList();
-        InputStream fileInputStream = null;
+    private List<FileSystemLocationFingerprint> fingerprintZipEntries(String zipFile) throws IOException {
+        ZipInput input = null;
         try {
-            fileInputStream = Files.newInputStream(Paths.get(jarFile));
-            ZipInputStream zipInput = new ZipInputStream(fileInputStream);
-            ZipEntry zipEntry;
-            RelativePathFactory relativePathFactory = new RelativePathFactory();
-
-            while ((zipEntry = zipInput.getNextEntry()) != null) {
-                relativePathFactory.setZipEntry(zipEntry);
-                if (zipEntry.isDirectory() || classpathResourceFilter.shouldBeIgnored(relativePathFactory)) {
-                    continue;
-                }
-                HashCode hash = classpathResourceHasher.hash(zipEntry, zipInput);
+            input = ZipInput.create(new File(zipFile));
+            return fingerprintZipEntries(input);
+        } finally {
+            IoActions.closeQuietly(input);
+        }
+    }
+
+    private List<FileSystemLocationFingerprint> fingerprintZipEntries(ZipInput input) throws IOException {
```

Comment:
It seems to me we're creating and merging lots of ArrayLists here.
Also, we're copying the whole thing over at the end, just to sort it.
Could we collect everything to a TreeSet instead, like:
```java
private void fingerprintZipEntries(ZipInput input, TreeSet<FileSystemLocationFingerprint> result) throws IOException {
    ZipEntry zipEntry;
    RelativePathFactory relativePathFactory = new RelativePathFactory();
    while ((zipEntry = input.getNextEntry()) != null) {
        relativePathFactory.setZipEntry(zipEntry);
        if (!zipEntry.isDirectory() && !resourceFilter.shouldBeIgnored(relativePathFactory)) {
            if (isZipFile(zipEntry.getName())) {
                fingerprintZipEntries(ZipInput.create(zipEntry.getInputStream()), result);
            } else {
                HashCode hash = resourceHasher.hash(zipEntry);
                if (hash != null) {
                    result.add(new DefaultFileSystemLocationFingerprint(zipEntry.getName(), FileType.RegularFile, hash));
                }
            }
        } 
    }
}
```
in which case we wouldn't need sorting at the end either:
```java
public static void appendSortedToHasher(Hasher hasher, TreeSet<FileSystemLocationFingerprint> sortedFingerprints) {
    for (FileSystemLocationFingerprint normalizedSnapshot : sortedFingerprints) {
        normalizedSnapshot.appendToHasher(hasher);
    }
}
```

What do you think @oehme, @lptr?

Reply:
I've created a simple test to see the speed before and after (it hashes every `.jar` file in my Gradle cache), and it shows a significant improvement for your changes: for 1780 archives it finishes in `20 seconds` before, and `16 seconds` after - 25% improvement, congrats!.
My suggested changes don't seem to significantly affect the end-result.

Reply:
Your experiment ignores that jar hashes are cached, so you'll only ever see these 20s on a machine with an empty cache. In that case those 20s are probably your smalles problem. This is why I keep saying you have to measure real builds, not synthetic benchmarks. 

We can still consider these changes, but let's not add more changes to this PR, which is about supporting more kinds of zips, not optimizing the algorithm.

Reply:
Ok, sure, I'll provide my ideas in separate pull requests.
Your PR already improves it a lot :)